### ISSUE TITLE
Cleaner API for HTTP headers

### ISSUE THREAD

**Issue by timboudreau** [netty/netty#3322]
I see the vast number of methods for getting various typed values out of `HttpHeaders` as dates, numbers, etc. (maybe these are new-ish?  I don't remember them from when I was writing Acteur).

I see what you're trying to do here, but it seems clunky.  A pattern similar to what you have for `ChannelOption<T>` would be cleaner and less bug-prone.  For example, a `Vary` header's value type ought to be a list of kinds of headers - and so forth.  And having the API make it difficult to an inappropriate type is a plus.

It happens that I have such an API in Acteur.  It wouldn't take many changes to port that to Netty's codebase - I'd just have to remove the dependency on Joda Time.  Typical code in Acteur looks like:

```java
add (Headers.IF_MODIFIED_SINCE, DateTime.now());
add(Headers.AGE, Duration.standardDays(2));
// and non-standard headers
add(Headers.stringHeader("X-foo"), "foo");
```

Here's [the package with that API](https://github.com/timboudreau/acteur/tree/master/acteur-headers/src/main/java/com/mastfrog/acteur/headers) and the [base type for Header types](https://github.com/timboudreau/acteur/blob/master/acteur-headers/src/main/java/com/mastfrog/acteur/headers/HeaderValueType.java) and the [collection of built-in headers](https://github.com/timboudreau/acteur/blob/master/acteur-headers/src/main/java/com/mastfrog/acteur/headers/Headers.java)

If you think this would be interesting, let me know.

**Comment by Scottmitch** [comment]
@timboudreau - This does sound interesting.  Just for some background I think there were 4 overarching goals when re-factoring the headers interfaces and class hierarchy:
1. Ensure all header interfaces were consistent (class hierarchy consolidation)
2. All header classes should share a common implementation (where applicable...they are mostly doing the same thing).
3. Improve the header algorithm which required full iteration (and possibly a collection reversal operation) for most operations.
4. Make it more apparent when "implicit" translations where being done.

Some more background on 4...It was very common place (and may still be) that conversions between AsciiString/String/CharSequence/Native Types were done that were not always necessary (or done externally to the headers interface after the fact in an inconsistent manner).  This may be less a responsibility of the interface and more a responsibility for the user, but since it was prevalent in Netty itself the assumption was the interface could be modified to help this.

As part of reviews I think another implicit goal was to support native types without necessarily requiring an additional boxing/unboxing.  This may not be necessary with your approach but can you clarify on this?

What may also need some thought for your approach to be merged in is that not all headers use a common "native base type".  For example we have [TextHeaders](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/TextHeaders.java) which use `CharSequence` and [BinaryHeaders](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/BinaryHeaders.java) which uses `AsciiString`.  So the [HeaderValueType](https://github.com/timboudreau/acteur/blob/master/acteur-headers/src/main/java/com/mastfrog/acteur/headers/HeaderValueType.java) may need a bit of work before merging in?

**Comment by Scottmitch** [comment]
@timboudreau - Also important to note is that the 4.0 branch HttpHeaders is a bit different than the 4.1 and 5.0 branches for backward compatibility reasons.

**Comment by timboudreau** [comment]
I already have a version of these classes which works with Netty 5 on the `netty-5` branch, for what that's worth.

**Comment by Scottmitch** [comment]
Would you mind submitting a PR?

**Comment by timboudreau** [comment]
If you speak for @trustin and @normanmaurer that you actually want it, then I'll put some time into it.  It will be a bit of work to rip out some stuff, so I want to be pretty sure the work would have a future before I do it.

**Comment by Scottmitch** [comment]
Makes sense before you invest more time to be sure everyone is on board.  However if you already have working code it may be easier to discuss if you submit a "work in progress" PR with what you have now.  For example a few of my questions from https://github.com/netty/netty/issues/3322#issuecomment-69580921 may be cleared up if the code already exists.

**Comment by timboudreau** [comment]
I have working code, but it has external dependencies on Joda Time and a
few other things I know you won't want.

Check out the sources for [acteur](https://github.com/timboudreau/acteur)
and look at the `acteur-headers` project - it's a fairly thin layer on top
of Netty, so that should give a pretty good idea of what it would look like.

On Mon, Jan 12, 2015 at 10:58 AM, Scott Mitchell <notifications@github.com>
wrote:

> Makes sense before you invest more time to be sure everyone is on board.
> However if you already have working code it may be easier to discuss if you
> submit a "work in progress" PR with what you have now. For example a few of
> my questions from #3322 (comment)
> <https://github.com/netty/netty/issues/3322#issuecomment-69580921> may be
> cleared up if the code already exists.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/netty/netty/issues/3322#issuecomment-69591845>.
>



-- 
http://timboudreau.com

**Comment by Scottmitch** [comment]
I took another look at your code.  Here are the notable differences I picked up on:
1. No native method support
2. Opportunity for more "direct" mapping to java objects
3. Conversions and value types are divided with class hierarchy as apposed to a conversion interface (i.e. [Headers.ValueConverter](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/Headers.java#L54)).

Here is a brief pros/cons summary:
`-` No native methods support - additional boxing/unboxing is required for native types.
`+` No native methods support - less interface to maintain.  Although I don't think the number of native types are going to be changing.

`+` More direct mapping to java objects - More compact and potentially simpler access to headers for users.
`-` More direct mapping to java objects - May implicitly do conversions or require extra iterations (for arrays) that may not necessary or desired

`+` Value Type hierarchy - Decoupling value types...separation of concerns.
`?` Value Type hierarchy - Easier to extend? Override existing `convertObject` vs some type of `HeaderValueType.toValue`.  This is likely more of a `+` than a `-`.
`-` Value Type hierarchy - Decoupling value types...centralized conversion may ensure more uniform conversions and less boiler plate class structure/files around the conversions.

In addition to these points I think the existing [HeaderValueType](https://github.com/timboudreau/acteur/blob/master/acteur-headers/src/main/java/com/mastfrog/acteur/headers/HeaderValueType.java) would need an additional generic type parameter to be used for [HeaderValueType.name()](https://github.com/timboudreau/acteur/blob/master/acteur-headers/src/main/java/com/mastfrog/acteur/headers/HeaderValueType.java#L45) and [HeaderValueType.toValue(...)](https://github.com/timboudreau/acteur/blob/master/acteur-headers/src/main/java/com/mastfrog/acteur/headers/HeaderValueType.java#L62).  Is this the approach you took?

In summary I see some benefits but it is not clear exactly what changes you are suggesting.  Can you please be more specific as to what changes you are proposing?  For example are you suggesting an alternative signature for all the methods that use the [Headers$ValueConverter.convertObject](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/Headers.java#L55)?

**Comment by timboudreau** [comment]
What exactly do you mean by "native methods support"?  You want a guarantee that a header set as, say, a primitive `long` will not at any point be converted to a `Long`?

If so, my thought is that, if the ability to do that is already there (I'm not suggesting deleting the support for doing that - we can have that conversation, but I don't want to tie the question of having a more straightforward API to nuking that), than someone fixated on microoptimization (and this is the textbook case of one) can use it.  But for mere mortals who benefit more from an API that uses types to make easy classes of mistakes impossible (if I set a value of `Set-Cookie to an `int`, what happens, exactly?  Most likely something goes boom on the other end of the wire, with their code none the wiser).  I get the desire to avoid boxing and memory copies if you're writing Twitter - but not everyone is.

`HeaderValueType` would need an additional parameter for the name?  In the Netty 5 branch, it is `CharSequence` - you want names to be able to be, say, `int`s?  I.e. so this stuff is reusable with some protocol where, under the hood, names are binary types of some sort?

For that I'd say, be wary of trying to come up with the One Header API To Rule Them All, because that dumps a lot of complexity on the API user who's just trying to figure out how to set `Last-Modified` and have it actually come out formatted right on the wire - they don't care what the type of header names is.  That said, you could simply enough do something like:

```java
interface HeaderValueType<NameType,ValueType> { ... }

interface HttpHeaderValueType<ValueType> extends HeaderValueType<AsciiString,ValueType> { ... }
```

**Comment by Scottmitch** [comment]
> What exactly do you mean by "native methods support"

Sorry, I have been a bit sloppy with terminology.  In this case primitive types.

> You want a guarantee that a header set as, say, a primitive long will not at any point be converted to a Long?

Not necessarily a guarantee, but at least that the interface doesn't preclude this opportunity.  Having an interface that exposes primitive types provides the opportunity to avoid converting to `Long` and just go directly to the `Header` underlying type (i.e. `CharSequence`).  This is not a necessity but since it is already there (and I don't foresee the primitive types changing) what is the rational for changing this?

> HeaderValueType would need an additional parameter for the name?

I'm not sure I understand what your asking because your demonstration interface `interface HeaderValueType<NameType,ValueType> { ... }` has additional generic parameter `NameType` (your [HeaderValueType](https://github.com/timboudreau/acteur/blob/master/acteur-headers/src/main/java/com/mastfrog/acteur/headers/HeaderValueType.java) only has 1 generic parameter).  Can you clarify your question?

> In the Netty 5 branch, it is CharSequence

Not necessarily true.  [TextHeaders](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/TextHeaders.java) uses CharSequence but [BinaryHeaders](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/BinaryHeaders.java) uses AsciiString (which is a CharSequence but is used for its binary storage. BinaryHeaders may change to something more strictly "binary" in the future).

> - you want names to be able to be, say, ints?

Not primitive types but the [Headers](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/Headers.java) class hierarchy currently supports a generic (non-primitive) types for the names.

> I.e. so this stuff is reusable with some protocol where, under the hood, names are binary types of some sort?

HTTP/2 uses "binary" headers.

> For that I'd say, be wary of trying to come up with the One Header API To Rule Them All

I gave some background on the goals for the headers re-factor here https://github.com/netty/netty/issues/3322#issuecomment-69580921.  The current class hierarchy is such that all "header" types support the same set of methods, and share a common implementation.  Before the re-factor this was not the case (class hierarchy was not uniform and there were multiple implementations).  This was done for some of the reasons you describe (simply/unify API, reduce bugs by consolidating implementation, etc...) and the resulted in [Headers](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/Headers.java).  Based upon the issue title (`Cleaner API for HTTP headers`) and your description (`HttpHeaderValueType<..>`) it seems like you want to break this paradigm and modify only the [HttpHeaders](https://github.com/netty/netty/blob/master/codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaders.java) interface?  I want to better understand if this is what you are suggesting and what the rational for not keeping the existing uniform interface is.  For example it seems like the `HeaderValueType` can be made a bit more generic to potentially keep inline with these goals and insert lower (closer to `Headers`) than the `HttpHeaders` interface in the hierarchy.

> because that dumps a lot of complexity on the API user

I don't think that is necessarily true in this case.  I'm not sure the current interface is more complicated than the previous set of interfaces from a user perspective.  It has the added benefit of being consistent over all netty codecs and so if you know how to use it for 1 codec then the learning curve is not as steep for other codecs.  That is not to say the existing interface cannot be improved, but that is why I'm trying to understand your suggestion :)

> they don't care what the type of header names is.

Can you expand upon this point?  I'm not sure I see how your suggestion would reduce what the users needs to know about type for the header name.  Seems like they would be required to know additional types (i.e. `HttpHeaderValueType`).
